source/common/config/config_provider_impl.h

```diff
@@ -184,25 +190,48 @@ class ConfigSubscriptionCommonBase : protected Logger::Loggable<Logger::Id::conf
     init_target_.ready();
   }
 
-  /**
-   * Returns one of the bound mutable config providers.
-   * @return const MutableConfigProviderCommonBase* a const pointer to a
-   *         bound MutableConfigProviderCommonBase or nullptr when there are none.
-   */
-  MutableConfigProviderCommonBase* getAnyBoundMutableConfigProvider() const {
-    return !mutable_config_providers_.empty() ? *mutable_config_providers_.begin() : nullptr;
-  }
-
 protected:
+  struct ThreadLocalConfig : public ThreadLocal::ThreadLocalObject {
+    explicit ThreadLocalConfig(ConfigProvider::ConfigConstSharedPtr initial_config)
+        : config_(std::move(initial_config)) {}
+
+    ConfigProvider::ConfigConstSharedPtr config_;
+  };
+
   ConfigSubscriptionCommonBase(const std::string& name, const uint64_t manager_identifier,
                                ConfigProviderManagerImplBase& config_provider_manager,
-                               TimeSource& time_source, const SystemTime& last_updated,
-                               const LocalInfo::LocalInfo& local_info)
-      : name_(name),
+                               Server::Configuration::FactoryContext& factory_context)
+      : name_(name), tls_(factory_context.threadLocal().allocateSlot()),
         init_target_(fmt::format("ConfigSubscriptionCommonBase {}", name_), [this]() { start(); }),
         manager_identifier_(manager_identifier), config_provider_manager_(config_provider_manager),
-        time_source_(time_source), last_updated_(last_updated) {
-    Envoy::Config::Utility::checkLocalInfo(name, local_info);
+        time_source_(factory_context.timeSource()),
+        last_updated_(factory_context.timeSource().systemTime()) {
+    Envoy::Config::Utility::checkLocalInfo(name, factory_context.localInfo());
+  }
+
+  /**
+   * Propagates a config update to worker threads.
+   *
+   * @param update_fn the callback to run on each thread, it takes the previous version Config and
+   * returns a updated/new version Config.
+   * @param complete_cb the callback to run when the update propagation is done.
+   */
+  void applyConfigUpdate(
+      const std::function<
+          ConfigProvider::ConfigConstSharedPtr(ConfigProvider::ConfigConstSharedPtr)>& update_fn,
```

Comment:
Can you use typedef/using aliases for these? It helps make clearer what the function types are really meaning.

Reply:
done